### PR TITLE
Fixed whisper.cpp implementation in setup script

### PR DESCRIPTION
**Pull request by ba2512005**
Modifications on setup script. Run with sudo -E env "PATH=$PATH" ./setup.sh to preserve ENV vars. 

Added model picker for whisper, default to medium
Automated setup script, auto download models
Auto detect Nvidia capable GPU, make with cublas
If not, make with cpu only.

Currently having trouble with the start.sh script:
basharat@escapepod:~/wire-pod$ sudop ./chipper/start.sh 
++ uname -a
+ UNAME='Linux escapepod 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux'
+ [[ 0 -ne 0 ]]
+ [[ -d ./chipper ]]
+ cd chipper
+ [[ ! -f ./source.sh ]]
+ source source.sh
++ export DEBUG_LOGGING=true
++ DEBUG_LOGGING=true
++ export STT_SERVICE=whisper.cpp
++ STT_SERVICE=whisper.cpp
++ export STT_MODEL=medium
++ STT_MODEL=medium
+ [[ whisper.cpp == \l\e\o\p\a\r\d ]]
+ [[ whisper.cpp == \r\h\i\n\o ]]
+ [[ whisper.cpp == \h\o\u\n\d\i\f\y ]]
+ [[ whisper.cpp == \w\h\i\s\p\e\r ]]
+ [[ whisper.cpp == \w\h\i\s\p\e\r\.\c\p\p ]]
+ [[ -f ./chipper ]]
+ export C_INCLUDE_PATH=/home/basharat/wire-pod/whisper.cpp
+ C_INCLUDE_PATH=/home/basharat/wire-pod/whisper.cpp
+ export LIBRARY_PATH=/home/basharat/wire-pod/whisper.cpp
+ LIBRARY_PATH=/home/basharat/wire-pod/whisper.cpp
+ export CGO_CFLAGS=-I/wire-pod/whisper.cpp/
+ CGO_CFLAGS=-I/wire-pod/whisper.cpp/
+ export CGO_LDFLAGS=-L/wire-pod/whisper.cpp/
+ CGO_LDFLAGS=-L/wire-pod/whisper.cpp/
+ [[ Linux escapepod 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux == *\D\a\r\w\i\n* ]]
+ C_INCLUDE_PATH=/home/basharat/wire-pod/whisper.cpp
+ LIBRARY_PATH=/home/basharat/wire-pod/whisper.cpp
+ /usr/local/go/bin/go run cmd/experimental/whisper.cpp/main.go
go: downloading github.com/wlynxg/anet v0.0.1
go: downloading github.com/kercre123/zeroconf v1.0.1
go: downloading github.com/miekg/dns v1.1.41
go: downloading github.com/cenkalti/backoff v2.2.1+incompatible
# command-line-arguments
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: /home/basharat/wire-pod/whisper.cpp/libwhisper.a(whisper.o): in function `whisper_init_state':
whisper.cpp:(.text+0xc7ad): undefined reference to `ggml_cublas_loaded'
/usr/bin/ld: whisper.cpp:(.text+0xd0d5): undefined reference to `ggml_backend_cuda_init'
/usr/bin/ld: /home/basharat/wire-pod/whisper.cpp/libwhisper.a(whisper.o): in function `whisper_model_load(whisper_model_loader*, whisper_context&)':
whisper.cpp:(.text+0x19d3f): undefined reference to `ggml_cublas_loaded'
/usr/bin/ld: whisper.cpp:(.text+0x19d6b): undefined reference to `ggml_backend_cuda_init'
collect2: error: ld returned 1 exit status

### CODE REVIEW

File: chipper/start.sh [hunk 1]
```diff
@@ -58,13 +59,17 @@ elif [[ ${STT_SERVICE} == "whisper.cpp" ]]; then
     export LIBRARY_PATH="../whisper.cpp"
     ./chipper
   else
-    export C_INCLUDE_PATH="../whisper.cpp"
-    export LIBRARY_PATH="../whisper.cpp"
+    export C_INCLUDE_PATH=~/wire-pod/whisper.cpp
```
Comment (xiJulian) [inline chipper/start.sh:62]: wire-pod isn't necessarily located in this path

**Review by xiJulian — CHANGES_REQUESTED** [review]

File: chipper/pkg/wirepod/stt/whisper.cpp/WhisperCpp.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ func Init() error {
 		sttLanguage = strings.Split(vars.APIConfig.STT.Language, "-")[0]
 	}
 
-	modelPath := filepath.Join(vars.WhisperModelPath, "ggml-tiny.bin")
+	modelPath := filepath.Join(vars.WhisperModelPath, "ggml-medium.bin")
```
[the remark this answers is not in the document]
Comment (ba2512005) [reply chipper/pkg/wirepod/stt/whisper.cpp/WhisperCpp.go:32]: When I was testing it, it wasn't as accurate. I can re-test it with the gpu. maybe the base model might be a good middle ground.

File: setup.sh [hunk 3]
```diff
@@ -140,12 +148,11 @@ function getSTT() {
...
-        echo "3: VOSK (local, accurate, multilanguage, fast, recommended)"
-        # echo "4: Whisper (local, accurate, multilanguage, a little slower, recommended for more powerful hardware)"
-        echo
-        read -p "Enter a number (3): " sttServiceNum
+        echo "3: VOSK (local, inaccurate, multilanguage, fast)"
+        echo "4: Whisper (cpp implementation, medium model, local, accurate, multilanguage, a little slower, recommended for more powerful hardware)"
+        read -p "Enter a number (4): " sttServiceNum
         if [[ ! -n ${sttServiceNum} ]]; then
-            sttService="vosk"
+            sttService="whisper.cpp"
```
[the remark this answers is not in the document]
Comment (ba2512005) [reply setup.sh:155]: Vosk is terrible for accuracy in my experience. Whisper is far superior. We can keep the Vosk default for now though to support more hardware

File: setup.sh [hunk 4]
```diff
@@ -157,20 +164,15 @@ function getSTT() {
...
             sttServicePrompt
         fi
     }
-    if [[ "$STT" == "vosk" ]]; then
-        echo "Vosk config"
-        sttService="vosk"
-    else
-        sttServicePrompt
-    fi
+    sttServicePrompt
```
[the remark this answers is not in the document]
Comment (ba2512005) [reply setup.sh:175]: This calls the function.

File: chipper/start.sh (hunk 1 above)
Comment (ba2512005) [reply chipper/start.sh:62]: There's a relative import above it, this line is redundant. I'm still modifying this file to fine tune it.

File: setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiJulian) [reply setup.sh:155]: the thing is that wire-pod is usually run on low-end hardware which would take more than three seconds to process a command compared to less than a second with vosk (it was tested by Wire)

File: setup.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (xiJulian) [reply setup.sh:175]: i think you accidentally deleted the lines above it

File: setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (kercre123) [reply setup.sh:155]: The majority of wire-pod instances are being run on Android T.V. boxes which have CPUs akin to what is in a Raspberry Pi 4. That is what wire-pod should be targeting IMO. Coqui, Picovoice, and VOSK run nicely on the Pi... whisper does not.

I wouldn't mind if Whisper is an option, but it should not be the default.

There are plenty of other VOSK models to try if you are having trouble with the default English (US) model. Accuracy has been great for me, even with the smaller 40MB model, but I have a very American voice. [https://alphacephei.com/vosk/models](https://alphacephei.com/vosk/models)

File: chipper/pkg/wirepod/stt/whisper.cpp/WhisperCpp.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kercre123) [reply chipper/pkg/wirepod/stt/whisper.cpp/WhisperCpp.go:32]: The tiny model should be plenty accurate for what Vector needs to understand.

File: setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (ba2512005) [reply setup.sh:155]: I understand, I couldn't get the whisper.cpp to work anyways. may as well just scrap this PR

**Comment by kercre123** [comment]
I uncommented the whisper.cpp setup code already in setup.sh and it worked just fine. I'll further implement it and include instructions on the wiki for setting it up with an NVIDIA card.

**Pull request closed without merging**